src/mtcp/mtcp_restart.c

```diff
@@ -519,20 +523,47 @@ setupTempRegions() {
     mtcp_abort();
   }
 
+  // In May 2020, on Cori and elsewhere, vvar is 3 pages and vdso is 2 pages.
   while (mtcp_readmapsline(mapsfd, &area)) {
     if (prev_addr + 5 * PAGESIZE <= area.addr) {
       vvarStartTmp = prev_addr;
       vdsoStartTmp = prev_addr + 3 * PAGESIZE;
       break;
     }
-    prev_addr = ROUNDADDRUP((uint64_t) area.endAddr, 0x1000);
+    prev_addr = ROUNDADDRUP((uint64_t) area.endAddr, PAGESIZE);
   }
   mtcp_sys_close(mapsfd);
 
   if (vvarStartTmp < 0 || vdsoStartTmp < 0) {
     MTCP_PRINTF("No free region found to temporarily map vvar/vdso to\n");
     mtcp_abort();
   }
+
+  // We have to iterate through the memory map twice, this time to ensure that
+  // free regions have been found for us to remap to.
+  mapsfd = mtcp_sys_open2("/proc/self/maps", O_RDONLY);
+  if (mapsfd < 0) {
+    MTCP_PRINTF("error opening /proc/self/maps; errno: %d\n", mtcp_sys_errno);
+    mtcp_abort();
+  }
+
+  while (mtcp_readmapsline(mapsfd, &area)) {
+    void *rc = 0x0;
+    if (mtcp_strcmp(area.name, "[vvar]") == 0) {
+      rc = mtcp_sys_mremap(area.addr, area.size, area.size,
+                           MREMAP_FIXED | MREMAP_MAYMOVE, vvarStartTmp);
+    }
+    if (mtcp_strcmp(area.name, "[vdso]") == 0) {
+      rc = mtcp_sys_mremap(area.addr, area.size, area.size,
+                           MREMAP_FIXED | MREMAP_MAYMOVE, vdsoStartTmp);
+    }
+    if (rc == MAP_FAILED) {
+      MTCP_PRINTF("mtcp_restart failed: "
+                  "gdb --mpi \"DMTCP_ROOT/bin/mtcp_restart\" to debug.\n");
```

Comment:
do we want to abort to return an error here?

Reply:
yes, i think we should, but i would defer to the opinions of @gc00 and @JainTwinkle.